### PR TITLE
Contributing lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py module to manage HUAWEI data center CloudEngine

### PR DESCRIPTION
**Pull request by QijunPan**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

ce_snmp_target_host.py
##### ANSIBLE VERSION

```
ansible 2.3.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/ansible-2.1.1.0/lib/ansible/modules/core/network/cloudengine', '/home/ansible-2.1.1.0/lib/ansible/module_utils', '/home/ansible-2.1.1.0/lib/ansible/modules', '/home/ansible-2.1.1.0', '/home/ansible-2.1.1.0/lib/ansible/utils/']
```

##### SUMMARY

```
TASK [change snmp target host version] *****************************************
task path: /usr1/code/OpenNESS/code/current/KVM/intg/test/testcases/test-ce_snmp_target_host.yml:31
Using module file /usr/local/lib/python2.7/site-packages/ansible-2.3.0-py2.7.egg/ansible/modules/core/network/cloudengine/ce_snmp_target_host.py
<ce_6850> ESTABLISH LOCAL CONNECTION FOR USER: root
<ce_6850> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1487053689.97-5279664452771 `" && echo ansible-tmp-1487053689.97-5279664452771="` echo $HOME/.ansible/tmp/ansible-tmp-1487053689.97-5279664452771 `" ) && sleep 0'
<ce_6850> PUT /tmp/tmpMAwtn5 TO /root/.ansible/tmp/ansible-tmp-1487053689.97-5279664452771/ce_snmp_target_host.py
<ce_6850> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1487053689.97-5279664452771/ /root/.ansible/tmp/ansible-tmp-1487053689.97-5279664452771/ce_snmp_target_host.py && sleep 0'
<ce_6850> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1487053689.97-5279664452771/ce_snmp_target_host.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1487053689.97-5279664452771/" > /dev/null 2>&1 && sleep 0'
changed: [ce_6850] => {
    "changed": true, 
    "end_state": {
        "version": "v3"
    }, 
    "existing": {
        "version": "v2c"
    }, 
    "invocation": {
        "module_args": {
            "address": null, 
            "auth_pass": null, 
            "authorize": false, 
            "connect_port": null, 
            "host": "ce_6850", 
            "host_name": null, 
            "interface_name": null, 
            "is_public_net": null, 
            "notify_type": null, 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 12345, 
            "provider": null, 
            "recv_port": null, 
            "security_level": null, 
            "security_model": null, 
            "security_name": null, 
            "security_name_v3": null, 
            "ssh_keyfile": null, 
            "state": "present", 
            "timeout": 10, 
            "transport": null, 
            "use_ssl": false, 
            "username": "rootDC", 
            "validate_certs": true, 
            "version": "v3", 
            "vpn_name": null
        }, 
        "module_name": "ce_snmp_target_host"
    }, 
    "proposed": {
        "state": "present", 
        "version": "v3"
    }, 
    "updates": [
        "snmp-agent sys-info version v3"
    ]
}
```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py [hunk 1]
```diff
@@ -0,0 +1,947 @@
...
+        delete_flag = False
+        result["target_host_info"] = []
+
+        if self.host_name:
+
+            if len(self.host_name) > 32 or len(self.host_name) < 1:
+                self.module.fail_json(
+                    msg='Error: The len of host_name is out of [1 - 32].')
+
+            if self.state == "present":
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
required_together
```

File: lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+                if not self.address or not self.notify_type:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
required_together
```

File: lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py [hunk 3]
```diff
@@ -0,0 +1,947 @@
...
+                    self.module.fail_json(
+                        msg='Error: Please input address notify_type at the same time.')
+
+            if self.security_model and self.security_model in ["v1", "v2c"]:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
required_if
```

File: lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py [hunk 4]
```diff
@@ -0,0 +1,947 @@
...
+                if not self.security_name:
+                    self.module.fail_json(
+                        msg='Error: Please input security_name when use security_model [%s].' % self.security_model)
+
+            if self.security_model and self.security_model in ["v3"]:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
required_if
```

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py (hunk 4 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py]: thank you, The problem has been fixed.

File: lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py (hunk 3 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py]: thank you, The problem has been fixed.

File: lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py (hunk 2 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py]: thank you, The problem has been fixed.

File: lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py (hunk 1 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_target_host.py]: thank you, The problem has been fixed.

**Comment by QijunPan** [comment]
@gundalow   These problems have been fixed， can  this PR merged?

**Review by gundalow — APPROVED** [review]

**Pull request merged**